entity-framework/core/miscellaneous/events/interception.md

```diff
@@ -0,0 +1,750 @@
+
+SaveChanges can be intercepted to create an independent audit record of the changes made.
+
+> [!NOTE]
+> This is not intended to be a robust auditing solution. Rather it is a simplistic example used to demonstrate the features of interception.
+
+#### The application context
+
+The [sample for auditing](https://github.com/dotnet/EntityFramework.Docs/tree/master/samples/core/Miscellaneous/SaveChangesInterception) uses a simple DbContext with blogs and posts.
+
+<!--
+public class BlogsContext : DbContext
+{
+    private readonly AuditingInterceptor _auditingInterceptor = new AuditingInterceptor("DataSource=audit.db");
+
+    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
+        => optionsBuilder
+            .AddInterceptors(_auditingInterceptor)
+            .UseSqlite("DataSource=blogs.db");
+
+    public DbSet<Blog> Blogs { get; set; }
+}
+
+public class Blog
+{
+    public int Id { get; set; }
+    public string Name { get; set; }
+
+    public ICollection<Post> Posts { get; } = new List<Post>();
+}
+
+public class Post
+{
+    public int Id { get; set; }
+    public string Title { get; set; }
+
+    public Blog Blog { get; set; }
+}
+-->
+[!code-csharp[BlogsContext](../../../../samples/core/Miscellaneous/SaveChangesInterception/BlogsContext.cs?name=BlogsContext)]
+
+Notice that a new instance of the interceptor is registered for each DbContext instance. This is because the auditing interceptor contains state linked to the current context instance.
+
+#### The audit context
+
+The sample also contains a second DbContext and model used for the auditing database.
+
+<!--
+public class AuditContext : DbContext
+{
+    private readonly string _connectionString;
+
+    public AuditContext(string connectionString)
+    {
+        _connectionString = connectionString;
+    }
+
+    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
+        => optionsBuilder.UseSqlite(_connectionString);
+
+    public DbSet<SaveChangesAudit> SaveChangesAudits { get; set; }
+}
+
+public class SaveChangesAudit
+{
+    public int Id { get; set; }
+    public Guid AuditId { get; set; }
+    public DateTime StartTime { get; set; }
+    public DateTime EndTime { get; set; }
+    public bool Succeeded { get; set; }
+    public string ErrorMessage { get; set; }
+
+    public ICollection<EntityAudit> Entities { get; } = new List<EntityAudit>();
+}
+
+public class EntityAudit
+{
+    public int Id { get; set; }
+    public EntityState State { get; set; }
+    public string AuditMessage { get; set; }
+
+    public SaveChangesAudit SaveChangesAudit { get; set; }
+}
+-->
+[!code-csharp[AuditContext](../../../../samples/core/Miscellaneous/SaveChangesInterception/AuditContext.cs?name=AuditContext)]
+
+#### The interceptor
+
+The general idea for auditing with the interceptor is:
+
+* An audit message is created at the beginning of SaveChanges and is written to the auditing database
```

Comment:
Just out of curiosity: this makes two database audit operations for each operation (one to insert before, one to update after). Any reason not to just do one after (unless you're trying to show both methods :smile:)

Reply:
Because I want there to be a record even if the power goes out between the first and the second. But it doesn't matter much; it's just to demonstrate interceptors.